### PR TITLE
Make identification of climatological bounds / multi-year mean more flexible

### LINKED ISSUE #22: Make identification of climatological bounds / multi-year mean more flexible
Currently: 

- A multi-year mean file is identified by having a climatology bounds variable, identified on the time variable by the attribute 'climatology', that exists

Improvements:

- Add a strict-mode flag to relevant functions (climatology_bounds_var_name, is_multi_year_mean, others?) that defaults to False (non-strict)
- In strict mode: use current rule
- In non-strict mode (in order of likelihood of correctness): 
  - _climatology bounds variable_ is identified by the following heuristics:
    - time variable has 'climatology' attribute (current case): use it
    - variable with name 'climatology_bounds', 'climatology_bnds', 'climo_bounds', 'climo_bnds' exists, even if it not identified by `time:climatology`
    - time variable has 'bounds' (not 'climatology') attribute identifying an existing variable AND that time bounds variable brackets multi-year periods (corresponding to the climatological averaging)
     - variable with name 'time_bounds' or 'time_bnds' exists (not identified by time:bounds) AND that time bounds variable brackets multi-year periods (corresponding to the climatological averaging)
  - _multi-year mean_ is identified by the following heuristics:
    - climatology bounds variable identified and exists as above
    - time variable has "suspicious" length: 1, 4, 12, 5, 13, 16, 17 (yearly, seasonal, monthly, and various concatenations) AND time variable has likely values (mid-month, mid-season, mid-year)

### PR DESCRIPTION
**Pull request by rod-glover**
Resolves #22 
Addresses https://github.com/pacificclimate/climate-explorer-backend/issues/52

Testing these heuristics thoroughly is a lot of work but was worth it -- caught a corner case or two that would have escaped even relatively intense code reading.

### CODE REVIEW

**Comment by rod-glover** [comment]
The `strict` flag needs to be propagated into any props/methods that depend on `climatology_bounds_var_name`, `is_multi_year_mean`

File: nchelpers/__init__.py [hunk 1]
```diff
@@ -362,26 +396,135 @@ def reduced_dims(self, var_name=None):
...
+        """Return True if the metadata indicates that the data consists of a multi-year mean,
         i.e., if the file contains a climatological time bounds variable.
+
         See http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/
         cf-conventions.html#climatological-statistics,
-        section 7.4"""
-        return bool(self.climatology_bounds_var_name)
+        section 7.4
+
+        In non-strice mode, failing metadata that conforms to the above standard, we use heuristics.
```
Comment (jameshiebert) [inline nchelpers/__init__.py]: s/strice/strict/

File: nchelpers/__init__.py [hunk 2]
```diff
@@ -362,26 +396,135 @@ def reduced_dims(self, var_name=None):
...
+        These heuristics are emobodied here and in `get_climatology_bounds_var_name` in non-strict mode.
+
+        If `self._cf_dataset_options['strict_metadata']` is True, use strict rules only for determining if this file contains
+        multi-year means. Otherwise, use heuristics as well.
+        """
+        # If there is no time axis, this can't be a file of temporal means.
```
Comment (jameshiebert) [inline nchelpers/__init__.py]: Sensible :)

File: nchelpers/__init__.py [hunk 3]
```diff
@@ -362,26 +396,135 @@ def reduced_dims(self, var_name=None):
...
+                1: (check_yearly,),
+                4: (check_seasonal,),
+                12: (check_monthly,),
+                5: (check_seasonal, check_yearly),
+                13: (check_monthly, check_yearly),
+                16: (check_monthly, check_seasonal),
+                17: (check_monthly, check_seasonal, check_yearly),
+            }[time_var.size]
+        except KeyError:
+            pass  # Try next heuristic
```
Comment (jameshiebert) [inline nchelpers/__init__.py]: i.e. No suspicious lengths?

File: tests/conftest.py [hunk 4]
```diff
@@ -1,7 +1,15 @@
+import sys
+import os
+
 from pytest import fixture
 from pkg_resources import resource_filename
 from nchelpers import CFDataset
 
+# Add helpers directory to pythonpath: See https://stackoverflow.com/a/33515264
```
Comment (jameshiebert) [inline tests/conftest.py]: Can't you use relative imports to do this?

File: tests/test_CFDataset.py [hunk 5]
```diff
@@ -106,6 +112,154 @@ def test_simple_property(tiny_dataset, prop, expected):
...
+# Use a non-"suspcicious" number of time values to prevent false positives from that heuristic.
+narrow_time_bounds = [[0, 10], [10, 20]]
+narrow_time_values = [5, 15]
+
+wide_time_bounds = [[0, 3650], [30, 3680]]
+wide_time_values = [1825, 1855]
+
+# Starred components in lists would make this list construction much tidier, but Py <3.5 doesn't support that.
+climo_bounds_var_test_cases = (
+    # # Without time variable
```
Comment (jameshiebert) [inline tests/test_CFDataset.py]: Sorry, why is all of this commented out?

File: tests/conftest.py (hunk 4 above)
Comment (rod-glover) [inline tests/conftest.py]: Good grief, relative imports are confusing. There are dozens of stackoverflow questions about it, and few good answers. This was one answer I found a while ago. But after some thought, I was able to set things up for relative imports in the test directory. Phew.

**Comment by rod-glover** [comment]
This thing is (very nearly) ready for prime time. 

The only heuristic I didn't include was one that would allow for the end-month (instead of mid-month) dates found in `cgcm.nc`. What do you think? Should it be allowed? There is a test for it that is currently marked `xfail` because AFAICT the file _is_ a multi-year mean. See discussion in https://github.com/pacificclimate/climate-explorer-backend/issues/52

**Comment by rod-glover** [comment]
Argh. I rebased this onto master after merging another PR. Hence the slew of apparently new but actually old commits. Sorry about that. Nothing changed. (Honest.)

**Comment by rod-glover** [comment]
Please see comment re. mym heuristic with end-month. I'd like to get this thing merged soon.

**Comment by jameshiebert** [comment]
> The only heuristic I didn't include was one that would allow for the end-month

Hmmm, yes those dates don't seem right (but it looks like they were generated by ``cdo ymonmean`` from the netcdf history?):
```
$ ncdump -t -v time cgcm.nc
...
time = "2039-01-31 12", "2039-02-28 12", "2039-03-31 12", "2039-04-30 12", 
    "2039-05-31 12", "2039-06-30 12", "2039-07-31 12", "2039-08-31 12", 
    "2039-09-30 12", "2039-10-31 12", "2039-11-30 12", "2039-12-31 12" ;
}
```
and the time bounds are just flat out wrong:
```
$ ncdump -t -v time_bnds cgcm.nc
...
time_bnds =
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02",
  "2010-01-01", "2010-01-02" ;
}
```

Maybe we shouldn't try to work around data that is obviously wrong...

**Comment by rod-glover** [comment]
Hah, I hadn't looked at the time bounds. I did notice history says it is the output of `cdo ymonmean`, which made me sympathetic. The [CDO manual](https://code.mpimet.mpg.de/projects/cdo/embedded/cdo.pdf) has this to say about `ymonmean`:

"The date information in an output field is the date of the last contributing input field."

So if we want to accommodate unmassaged output of `ymonmean`, we should allow for this.

In case we do, I tried running `ymonmean` on `tiny_gcm.nc` and noted the following:

- No time bounds variable was created.
- In this particular case, the last contributing time for January was 1975-01-01 12:00, which emphasizes that in these cases the time values can basically be any day in the month.

**Question: Do we want to accommodate unmassaged output of `ymonmean`?**

Unfortunately, quite a few test cases depend on this sketchy file, so regardless of the answer to this question I will have to modify it in some way to make it acceptable. But the answer will determine part of how I modify it.

**Comment by rod-glover** [comment]
I'm surprised that `ymonmean` doesn't create time bounds. There is not apparently any way to tell the period over which those means were formed from the raw output.

I am gravitating to a "no" answer to the question of accommodating raw ymonmean, but I will be happy to with "yes" if there is a good reason to do so.

**Comment by rod-glover** [comment]
Going with "no"; minimally modifying `cgcm.nc` to pass heuristics (mid-month dates, reasonable time bounds).

**Comment by rod-glover** [comment]
@jameshiebert agrees with this decision, authorizes merge.

**Pull request merged**